### PR TITLE
Existingboiler and spaceheating

### PR DESCRIPTION
**Pull request by NLaws**
Added classes for boiler, existingboiler, spaceheating in new job endpoint

### Please check if the PR fulfills these requirements
- [x] CHANGELOG.md is updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?
Feature: the ability to provide a boiler and existing boiler techs along with space heating load information

### What is the current behavior?
(You can also link to an open issue here)
At present, if a user wishes to model space heating loads or work with Boiler techs, they have to utilize older version of the REopt API, or interact directly with Julia package. (#305 , #366)

### What is the new behavior (if this is a feature change)?
Now, a user can include Boiler, ExistingBoiler, and SpaceHeatingInput keys as inputs in their scenario JSON and expect results from REopt where these techs are considered.

### Does this PR introduce a breaking change?
(What changes might users need to make in their application due to this PR?)
No breaking changes

### CODE REVIEW

**Comment by rathod-b** [comment]
Hi Nick, any thoughts on tests running successfully locally but failing when executed via Github? What branch of REopt.jl does Github run the tests against?

**Comment by rathod-b** [comment]
Rm `Financial -> existing_boiler_fuel_cost_escalation_pct` and `ExistingBoiler -> can_supply_steam_turbine` since they are not in REopt.jl # develop at the moment.

File: job/models.py [hunk 1]
```diff
@@ -2758,6 +2790,532 @@ class Message(BaseModel, models.Model):
...
+        'FlatLoad_16_7 '
+        'FlatLoad_16_5 '
+        'FlatLoad_8_7 '
+        'FlatLoad_8_5'
+    ))
+
+    annual_mmbtu = models.FloatField(
+        validators=[
+            MinValueValidator(1),
+            MaxValueValidator(100000000)
```
Comment (zolanaj) [inline job/models.py:3225]: update `MaxValueValidator `value to `MAX_BIG_NUMBER` here as well

**Review by zolanaj — CHANGES_REQUESTED** [review]
@rathod-b @NLaws thank you!  I have a small collection of changes and discussion points, and one bug fix.  

Happy to discuss these if you have any questions @rathod-b.

File: job/src/process_results.py [hunk 2]
```diff
@@ -55,4 +57,58 @@ def process_results(results: dict, run_uuid: str) -> None:
...
+    "lifecycle_emissions_tonnes_PM25",
+    "lifecycle_emissions_tonnes_NOx_bau",
+    "lifecycle_emissions_tonnes_PM25_bau",
+    "lifecycle_emissions_from_fuelburn_tonnes_NOx",
+    "lifecycle_emissions_from_fuelburn_tonnes_PM25",
+    "lifecycle_emissions_from_fuelburn_tonnes_NOx_bau",
+    "lifecycle_emissions_from_fuelburn_tonnes_PM25_bau",
+    "emissions_region",
+    "distance_to_emissions_region_meters"
+]
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/src/process_results.py:105]: Hi Alex, did you have a specific place in mind? What do you think about moving this function to `run_jump_model.py` either after we update time in results (line 96) or immediately before `process_results()` is called (line 127)?

File: job/validators.py [hunk 3]
```diff
@@ -347,6 +353,14 @@ def cross_clean_pv(pvmodel):
...
                     self.add_validation_error("ElectricUtility", "outage_end_time_step",
                                               f"Value is greater than the max allowable ({max_ts})")
+        
+        """
+        ExistingBoiler
+        """
+        if "ExistingBoiler" in self.models.keys():
+
+            # Clean fuel_cost_per_mmbtu to align with time series
+            self.clean_time_series("ExistingBoiler", "fuel_cost_per_mmbtu")
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/validators.py:363]: Correct. We can add that validation when Boiler makes its way into develop or master branches?

File: job/models.py [hunk 4]
```diff
@@ -2807,5 +3365,26 @@ def filter_none_and_empty_array(d:dict):
...
+
+'''
+If a scalar was provided where API expects a vector, extend it to 8760
+Upsampling handled in InputValidator.cross_clean
+'''
+def scalar_to_vector(vec:list):
+    if len(vec) == 1: # scalar length is provided
+        return vec * 8760
+    elif len(vec) == 12: # Monthly costs were provided
+        return vec * 730
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/models.py:3388]: Good point, I will implement and test `numpy.repeat(vec, days_per_month)` where `days_per_month = [31,28,31,30,31,30,31,31,30,31,30,31]`.

File: job/src/process_results.py [hunk 5]
```diff
@@ -55,4 +57,58 @@ def process_results(results: dict, run_uuid: str) -> None:
...
         WindOutputs.create(meta=meta, **results["Wind"]).save()
+    if "Boiler" in results.keys():
+        BoilerOutputs.create(meta=meta, **results["Boiler"]).save()
+    if "ExistingBoiler" in results.keys():
+        ExistingBoilerOutputs.create(meta=meta, **results["ExistingBoiler"]).save()
     # TODO process rest of results
+
+keys_to_skip = [
+    "lifecycle_emissions_reduction_CO2_pct",
+    "breakeven_cost_of_emissions_reduction_us_dollars_per_tCO2",
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/src/process_results.py:68]: We shouldnt need to skip these inputs in `keys_to_skip`. If a key is provided in input JSON but its not present in one of the input models, it is skipped by default. Key `Financial -> existing_boiler_fuel_cost_escalation_pct` is in newboiler branch so we can either add it then, or we can add it now but leave it as nullable, blankable without default values.

File: job/models.py (hunk 1 above)
Comment (rathod-b) [reply job/models.py]: addressed in https://github.com/NREL/REopt_API/pull/322/commits/8b9bde382331eb675af40fb1f85d1abf97d53d49

File: job/models.py [hunk 6]
```diff
@@ -2758,6 +2790,532 @@ class Message(BaseModel, models.Model):
...
+        null=True,
+        blank=True,
+        default=0.0,
+        help_text="Existing boiler system efficiency - conversion of fuel to usable heating thermal energy."
+    )
+
+    emissions_factor_lb_CO2_per_mmbtu = models.FloatField(
+        validators=[
+            MinValueValidator(0.0),
+            MaxValueValidator(1.0e9)
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/models.py]: addressed in https://github.com/NREL/REopt_API/pull/322/commits/8b9bde382331eb675af40fb1f85d1abf97d53d49

File: julia_src/Manifest.toml [hunk 7]
```diff
@@ -42,6 +42,18 @@ git-tree-sha1 = "4c10eee4af024676200bc7752e536f858c6b8f93"
 uuid = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply julia_src/Manifest.toml]: addressed in https://github.com/NREL/REopt_API/pull/322/commits/8b9bde382331eb675af40fb1f85d1abf97d53d49. let me know if there are further questions/comments.

File: job/test/test_job_endpoint.py [hunk 8]
```diff
@@ -108,6 +108,99 @@ def test_pv_and_battery_scenario(self):
...
+        r = json.loads(resp.content)
+        results = r["outputs"]
+
+        assert("ExistingBoiler" in results)
+        # assert("Boiler" in results)
+        self.assertAlmostEqual(results["ExistingBoiler"]["year_one_fuel_consumption_mmbtu"], 29897, places=-3)
+        self.assertAlmostEqual(results["ExistingBoiler"]["lifecycle_fuel_cost_after_tax"], 160072, places=-3)
+        self.assertAlmostEqual(results["ExistingBoiler"]["year_one_fuel_cost_before_tax"], 14948, places=-3)
+        # self.assertAlmostEqual(results["Boiler"]["year_one_fuel_consumption_mmbtu"], 0.0)
+
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/test/test_job_endpoint.py]: Removed this assertion for now. Addressed in https://github.com/NREL/REopt_API/pull/322/commits/8b9bde382331eb675af40fb1f85d1abf97d53d49

File: job/test/test_job_endpoint.py [hunk 9]
```diff
@@ -108,6 +108,99 @@ def test_pv_and_battery_scenario(self):
...
+                "annual_mmbtu": 29897.6
+            },
+            "ElectricTariff": {
+                "urdb_utility_name": "Pacific Gas & Electric Co",
+                "urdb_rate_name": "E-20 Maximum demand of (1000 KW or more) (Secondary)",
+                "urdb_label": "5e1676e95457a3f87673e3b0"
+            }
+        }
+
+        del scenario["SteamTurbine"]
```
[the remark this answers is not in the document]
Comment (rathod-b) [reply job/test/test_job_endpoint.py]: addressed in https://github.com/NREL/REopt_API/pull/322/commits/8b9bde382331eb675af40fb1f85d1abf97d53d49

[the remark this answers is not in the document]
Comment (rathod-b) [reply job/test/test_job_endpoint.py]: addressed in https://github.com/NREL/REopt_API/pull/322/commits/8b9bde382331eb675af40fb1f85d1abf97d53d49

**Comment by zolanaj** [comment]
Thanks for clarifying.  I think let's leave it as nullable/blankable and add a TODO to ensure that these perform as expected once all the other pieces are merged (assuming the emissions does get reported each run, and only the constraints are optional).

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by zolanaj** [comment]
I believe each element in days_per_month should be multiplied by 24 to achieve the correct outcome, but I agree with the idea here.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by zolanaj** [comment]
Agreed.  Thanks!

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: job/src/process_results.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zolanaj) [reply job/src/process_results.py:68]: Let's opt for the former - this and the other NewBoiler-related updates can wait until that technology is fully added.

File: job/validators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zolanaj) [reply job/validators.py:362]: Agreed - thanks!

File: job/models.py [hunk 10]
```diff
@@ -1811,6 +1842,7 @@ class PVOutputs(BaseModel, models.Model):
     )
     size_kw = models.FloatField(null=True, blank=True)
     lifecycle_om_cost_after_tax = models.FloatField(null=True, blank=True)
+    lifecycle_om_cost_after_tax_bau = models.FloatField(null=True, blank=True)
     lifecycle_om_cost_bau = models.FloatField(null=True, blank=True)
 #     station_latitude = models.FloatField(null=True, blank=True)
```
[the remark this answers is not in the document]
Comment (zolanaj) [reply job/models.py:1848]: Got it, thanks for catching.  I agree that we'll need to vet all the i/o but we probably need to get the other techs to catch up.  We can do this with a big unit test if it's well-designed,  Punt for now.

File: job/src/process_results.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zolanaj) [reply job/src/process_results.py]: Either of those locations is fine; we just need to make sure it's removed after everything is merged and it's easiest to do that if it's an easily accessible place.  If we add a TODO to remove it later, that is probably enough.

File: job/models.py [hunk 11]
```diff
@@ -2758,6 +2790,532 @@ class Message(BaseModel, models.Model):
...
+                error_messages["blended_doe_reference_percents"] = "Sum must = 1.0."
+
+        if self.doe_reference_name != "" or \
+                len(self.blended_doe_reference_names) > 0:
+            self.year = 2017  # the validator provides an "info" message regarding this)
+
+        if error_messages:
+            raise ValidationError(error_messages)
+        
+        pass
```
[the remark this answers is not in the document]
Comment (zolanaj) [reply job/models.py:3321]: I'm adding issue https://github.com/NREL/REopt_API/issues/348 to address this at a later time.

**Comment by zolanaj** [comment]
@rathod-b thanks for addressing and responding to my comments.  If "Boiler" is removed from the PR name, I think we can unmark as draft and then merge into develop - sound good?  Or are there any nits I missed?

**Comment by rathod-b** [comment]
> I believe each element in days_per_month should be multiplied by 24 to achieve the correct outcome, but I agree with the idea here. Get Outlook for Android<https://aka.ms/AAb9ysg>
> ________________________________ From: Bhavesh Rathod ***@***.***> Sent: Friday, July 22, 2022 10:55:25 AM To: NREL/REopt_API ***@***.***> Cc: Zolan, Alexander ***@***.***>; Review requested ***@***.***> Subject: Re: [NREL/REopt_API] Boiler, existingboiler and spaceheating (PR #322) CAUTION: This email originated from outside of NREL. Do not click links or open attachments unless you recognize the sender and know the content is safe. @rathod-b commented on this pull request.
> ________________________________ In job/models.py<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNREL%2FREopt_API%2Fpull%2F322%23discussion_r927734083&data=05%7C01%7Calexander.zolan%40nrel.gov%7Cc8746fa787a14e1184db08da6bf23607%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637940985298017713%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pT4CygzW8fzSi3l3TWEQqsUX6dl3W21Z6UlhL93OxOk%3D&reserved=0>:
> + elif len(vec) == 12: # Monthly costs were provided
> + return vec * 730 Good point, I will implement and test numpy.repeat(vec, days_per_month) where days_per_month = [31,28,31,30,31,30,31,31,30,31,30,31]. — Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNREL%2FREopt_API%2Fpull%2F322%23discussion_r927734083&data=05%7C01%7Calexander.zolan%40nrel.gov%7Cc8746fa787a14e1184db08da6bf23607%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637940985298017713%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pT4CygzW8fzSi3l3TWEQqsUX6dl3W21Z6UlhL93OxOk%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAC7MUG2IKISEFYZXEKZPJ3LVVKY53ANCNFSM5VCUGBTQ&data=05%7C01%7Calexander.zolan%40nrel.gov%7Cc8746fa787a14e1184db08da6bf23607%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637940985298017713%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=r%2BZ%2FQ1QuYO5R86P3pxumJgP56GUe0Tx6h8Gc0xpXD1Q%3D&reserved=0>. You are receiving this because your review was requested.Message ID: ***@***.***>

The 24 hours per day is being taken care of in return statement where we scale days/month to their appropriate # of hours
```python
def scalar_to_vector(vec:list):
    if len(vec) == 1: # scalar length is provided
        return vec * 8760
    elif len(vec) == 12: # Monthly costs were provided
        days_per_month = [31,28,31,30,31,30,31,31,30,31,30,31]
        return numpy.repeat(vec, [i * 24 for i in days_per_month]).tolist()
    else:
        return vec # the vector len was not 1, handle it elsewhere

**Comment by rathod-b** [comment]
> Thanks for clarifying. I think let's leave it as nullable/blankable and add a TODO to ensure that these perform as expected once all the other pieces are merged (assuming the emissions does get reported each run, and only the constraints are optional). Get Outlook for Android<https://aka.ms/AAb9ysg>
> ________________________________ From: Bhavesh Rathod ***@***.***> Sent: Friday, July 22, 2022 11:25:23 AM To: NREL/REopt_API ***@***.***> Cc: Zolan, Alexander ***@***.***>; Review requested ***@***.***> Subject: Re: [NREL/REopt_API] Boiler, existingboiler and spaceheating (PR #322) CAUTION: This email originated from outside of NREL. Do not click links or open attachments unless you recognize the sender and know the content is safe. @rathod-b commented on this pull request.
> ________________________________ In job/src/process_results.py<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNREL%2FREopt_API%2Fpull%2F322%23discussion_r927761853&data=05%7C01%7Calexander.zolan%40nrel.gov%7Cee1081a7204746f02c6808da6bf6658b%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637941003261070447%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=5IY6Zxd2BRSjCzn076O%2BGFqXG9QEoMGwiVLLKNF7nBw%3D&reserved=0>:
> # TODO process rest of results
> + +keys_to_skip = [ + "lifecycle_emissions_reduction_CO2_pct", + "breakeven_cost_of_emissions_reduction_us_dollars_per_tCO2", We shouldnt need to skip these inputs in keys_to_skip. If a key is provided in input JSON but its not present in one of the input models, it is skipped by default. Key Financial -> existing_boiler_fuel_cost_escalation_pct is in newboiler branch so we can either add it then, or we can add it now but leave it as nullable, blankable without default values. — Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNREL%2FREopt_API%2Fpull%2F322%23discussion_r927761853&data=05%7C01%7Calexander.zolan%40nrel.gov%7Cee1081a7204746f02c6808da6bf6658b%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637941003261070447%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=5IY6Zxd2BRSjCzn076O%2BGFqXG9QEoMGwiVLLKNF7nBw%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAC7MUGZDCUXTFA3SN4HUKJ3VVK4OHANCNFSM5VCUGBTQ&data=05%7C01%7Calexander.zolan%40nrel.gov%7Cee1081a7204746f02c6808da6bf6658b%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637941003261070447%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=wB%2FvR8nB9AmSSAD6FRONjHpaTNpQHt7aDG0nTlLqNYY%3D&reserved=0>. You are receiving this because your review was requested.Message ID: ***@***.***>

Sounds good, i will add a TODO

**Comment by rathod-b** [comment]
@zolanaj I updated REopt TOML files to point to v0.17.0. Apart from that is there anything else to check for this PR?

**Review by zolanaj — APPROVED** [review]

**Review by zolanaj — APPROVED** [review]

**Comment by zolanaj** [comment]
> @zolanaj I updated REopt TOML files to point to v0.17.0. Apart from that is there anything else to check for this PR?

I went through comments and code one last time and things look good to me.  If tests pass, let's merge.

File: job/src/run_jump_model.py [hunk 12]
```diff
@@ -89,9 +89,9 @@ def run_jump_model(run_uuid):
         t_start = time.time()
         julia_host = os.environ.get('JULIA_HOST', "julia")
         response = requests.post("http://" + julia_host + ":8081/reopt/", json=data)
-        if response.status_code == 500:
-            raise REoptFailedToStartError(task=name, message=results["error"], run_uuid=run_uuid, user_uuid=user_uuid)
         response_json = response.json()
+        if response.status_code == 500:
+            raise REoptFailedToStartError(task=name, message=response_json["error"], run_uuid=run_uuid, user_uuid=user_uuid)
```
Comment (zolanaj) [inline job/src/run_jump_model.py:94]: @rathod-b any reason for this switch?  It seems like the error check on `response` should happen before creating `response_json` as the original order had it

Comment (zolanaj) [reply job/src/run_jump_model.py:94]: Ah, I see it now - the response_json["error"] is in the correct format.

**Comment by zolanaj** [comment]
@rathod-b I believe we're caught up with develop now so this should be ready to merge.  Do you want to take a quick look at these updates?

@Bill-Becker are we still on a merge-to-develop embargo?  In either case I just want to put this in line for when we're ready.

**Comment by Bill-Becker** [comment]
@zolanaj We've already merged some non-deployment stuff into develop, and especially if this is all job/dev/v3 related additions, I think it's fine to just merge this. The only risk is that we add complexity to debugging a deployment issue with lots of other non-deployment changes too, but again because this is not the main URL/app, I think we are fine.

**Comment by zolanaj** [comment]
@Bill-Becker thanks.  Merging now.

**Pull request merged**